src/System.Private.CoreLib/src/System/Runtime/CompilerServices/jithelpers.cs

```diff
@@ -102,7 +102,17 @@ internal static StackCrawlMarkHandle GetStackCrawlMarkHandle(ref StackCrawlMark
         internal static ref byte GetRawData(this object obj) =>
             ref Unsafe.As<RawData>(obj).Data;
 
-        internal static ref byte GetRawSzArrayData(this Array array) =>
-            ref Unsafe.As<RawSzArrayData>(array).Data;
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static ref T GetRawSzArrayData<T>(this T[] array) =>
```

Comment:
I expect that this change will regress performance when `T` is reference type.

Reply:
This comment seems to be coming up  frequently in these sorts of PRs. Wouldn't it be better to improve JIT handling of these scenarios rather than to always push the concern up to the caller?

Reply:
Yes ... but it is not an easy fix.

Reply:
Ick. :(
It seems that would be fatal to this PR, no? When I inspected the codegen everything looked correct since things were being inlined, but if there's an unintended side effect this might not be a good idea.

Aside: wouldn't generic Unsafe.As / Unsafe.Add / etc. suffer from the same issue? We use them everywhere.

Reply:
`Unsafe.Add` would just be ptr increments? `Unsafe.As` not generally used for reference types?

Reply:
@benaadams Almost every call site I changed used Unsafe.As and/or Unsafe.Add in addition to JitHelpers.GetRawSzArrayData. For call sites like Span.ToArray it's not unreasonable to think a reference type T would be used.

Reply:
> Unsafe.As / Unsafe.Add / etc. suffer from the same issue?

They do not. This issue is caused by generics wrapping other generics. `Unsafe.*` are leafs.

Reply:
Would having some of it in non-generic help? (Guessing not)

e.g.
```csharp
internal static ref T GetRawSzArrayData<T>(this T[] array) =>
    ref Unsafe.As<byte, T>(ref GetRawSzArrayData(array));

private static ref byte GetRawSzArrayData(this Array array) =>
    ref Unsafe.As<RawSzArrayData>(array).Data;
```

Reply:
> It seems that would be fatal to this PR, no?

The way to make it work without perf hit would be to handcraft the IL from the VM side, similar to how `Unsafe` is done in CoreLib.

Reply:
Interestingly, it looks like the existing (non-generic) method is already implemented at the VM side. In theory all we'd have to do is change the managed signature, and as long as the method name is the same things should just work since the VM will hook it up correctly.

https://github.com/dotnet/coreclr/blob/b364ef6c7872d90d75063570f25495286b87acfd/src/vm/jitinterface.cpp#L7064